src/CacheClient.cpp

```diff
@@ -0,0 +1,125 @@
+{
+    return SlowCallback(user_data, "[CacheRead]");
+}
+
+static int SlowCallback_CacheWrite(void *user_data)
+{
+    return SlowCallback(user_data, "[CacheWrite]");
+}
+
+enum Operation
+{
+    kOperationRead,
+    kOperationWrite
+};
+
+static bool Invoke_REAPI_Cache_Client(const HashDigest& digest, StatCache *stat_cache, const FrozenArray<FrozenFileAndHash>& outputFiles, ThreadState* thread_state, Operation operation, const Frozen::DagNode* dagNode, Mutex* queue_lock)
+{
+    ProfilerScope profiler_scope("InvokeCacheMe", thread_state->m_ProfilerThreadId, outputFiles[0].m_Filename);
+
+    const char* reapi_raw = getenv(kENV_REAPI_CACHE_CLIENT);
+    if (reapi_raw == nullptr)
+        return false;
+
+    PathBuffer pathbuf;
+    PathInit(&pathbuf, reapi_raw);
+    char reapi[kMaxPathLength];
+    PathFormat(reapi, &pathbuf);
+
+    char buffer[5000];
+    int totalWritten = 0;
+    char digestString[kDigestStringSize];
+    DigestToString(digestString, digest);
+
+    const char* cmd = operation == kOperationRead ? "down" : "up";
+    totalWritten += snprintf(buffer, sizeof(buffer), "%s %s %s00000000000000000000000000000002", reapi, cmd, digestString);
+
+    for (auto &it : outputFiles)
+    {
+        PathBuffer output;
+        PathInit(&output, it.m_Filename);
+        MakeDirectoriesForFile(stat_cache, output);
+        totalWritten += snprintf(buffer+totalWritten, sizeof(buffer)-totalWritten, " \"%s\" ", it.m_Filename.Get());
+    }
+
+    SlowCallbackData slowCallbackData;
+    slowCallbackData.node_data = dagNode;
+    slowCallbackData.time_of_start = TimerGet();
+    slowCallbackData.queue_lock = queue_lock;
+    slowCallbackData.build_queue = thread_state->m_Queue;
+
+    Log(kDebug,"%s\n",buffer);
+    ExecResult result = ExecuteProcess(buffer, 0, nullptr, nullptr, thread_state->m_ThreadIndex, false, operation == kOperationRead ? SlowCallback_CacheRead : SlowCallback_CacheWrite , &slowCallbackData);
+
+    if (operation == Operation::kOperationRead)
+        for (auto &it : outputFiles)
+            StatCacheMarkDirty(stat_cache, it.m_Filename, it.m_FilenameHash);
+
+    return result.m_ReturnCode == 0;
+}
+
+bool CacheClient::AttemptRead(const Frozen::DagNode* dagNode, HashDigest signature, StatCache* stat_cache, Mutex* queue_lock, ThreadState* thread_state)
+{
+    return Invoke_REAPI_Cache_Client(signature, stat_cache, dagNode->m_OutputFiles, thread_state, Operation::kOperationRead, dagNode, queue_lock );
+}
+bool CacheClient::AttemptWrite(const Frozen::DagNode* dagNode, HashDigest signature, StatCache* stat_cache, Mutex* queue_lock, ThreadState* thread_state)
+{
+    return Invoke_REAPI_Cache_Client(signature, stat_cache, dagNode->m_OutputFiles, thread_state, Operation::kOperationWrite, dagNode, queue_lock );
+}
+
+void GetCachingBehaviourSettingsFromEnvironment(bool* attemptReads, bool* attemptWrites)
+{
+    *attemptReads = false;
+    *attemptWrites = false;
+
+    const char* server = getenv(kENV_CACHE_SERVER_ADDRESS);
```

Comment:
How is the server address passed to cacheme.exe? Does that also read the env variable?

Reply:
Yes, we just need to ensure that `CACHE_SERVER_ADDRESS` is set in its environment.